src/EnergyPlus/IndoorIceRink.cc

```diff
@@ -0,0 +1,2797 @@
+            }
+        }
+
+        if (BeginEnvrnFlag && MyEnvrnFlagGeneral) {
+            ZeroSourceSumHATsurf = 0.0;
+            QRadSysSrcAvg = 0.0;
+            LastQRadSysSrc = 0.0;
+            LastSysTimeElapsed = 0.0;
+            LastTimeStepSys = 0.0;
+            MyEnvrnFlagGeneral = false;
+        }
+        if (!BeginEnvrnFlag) MyEnvrnFlagGeneral = true;
+
+        if (SystemType == DirectSystem) {
+            if (BeginEnvrnFlag && MyEnvrnFlagDRink(SysNum)) {
+
+                DRink(SysNum).RefrigInletTemp = 0.0;
+                DRink(SysNum).RefrigOutletTemp = 0.0;
+                DRink(SysNum).RefrigMassFlow = 0.0;
+                DRink(SysNum).CoolPower = 0.0;
+                DRink(SysNum).CoolEnergy = 0.0;
+
+                if (!MyPlantScanFlagDRink(SysNum)) {
+                    if (DRink(SysNum).RefrigInNode > 0) {
+                        InitComponentNodes(0.0,
+                                           DRink(SysNum).MaxRefrigMassFlow,
+                                           DRink(SysNum).RefrigInNode,
+                                           DRink(SysNum).RefrigOutNode,
+                                           DRink(SysNum).RefrigLoopNum,
+                                           DRink(SysNum).RefrigLoopSide,
+                                           DRink(SysNum).RefrigBranchNum,
+                                           DRink(SysNum).RefrigCompNum);
+                    }
+                }
+                MyEnvrnFlagDRink(SysNum) = false;
+            }
+        }
+        if (!BeginEnvrnFlag && SystemType == DirectSystem) MyEnvrnFlagDRink(SysNum) = true;
+
+        if (SystemType == IndirectSystem) {
+            if (BeginEnvrnFlag && MyEnvrnFlagIRink(SysNum)) {
+
+                IRink(SysNum).RefrigInletTemp = 0.0;
+                IRink(SysNum).RefrigOutletTemp = 0.0;
+                IRink(SysNum).RefrigMassFlow = 0.0;
+                IRink(SysNum).CoolPower = 0.0;
+                IRink(SysNum).CoolEnergy = 0.0;
+
+                if (!MyPlantScanFlagIRink(SysNum)) {
+                    if (IRink(SysNum).RefrigInNode > 0) {
+                        InitComponentNodes(0.0,
+                                           IRink(SysNum).MaxRefrigMassFlow,
+                                           IRink(SysNum).RefrigInNode,
+                                           IRink(SysNum).RefrigOutNode,
+                                           IRink(SysNum).RefrigLoopNum,
+                                           IRink(SysNum).RefrigLoopSide,
+                                           IRink(SysNum).RefrigBranchNum,
+                                           IRink(SysNum).RefrigCompNum);
+                    }
+                }
+                MyEnvrnFlagIRink(SysNum) = false;
+            }
+        }
+        if (!BeginEnvrnFlag && SystemType == IndirectSystem) MyEnvrnFlagIRink(SysNum) = true;
+
+        if (BeginTimeStepFlag && FirstHVACIteration) {
```

Comment:
Begin time step looks suspicious here.  Did you know that the HVAC time step loop can step back and re-run the same iteration at a new time step length?  You probably need to check whether the current time has advanced to a new simulation time and only then step forward (and if it hasn't then you need to clear out the current results and start anew)